File: 158__Read_N_Characters_Given_Read4_II-Call_multiple_times.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution(object):

    def __init__(self):
        self.buf4 = [None] * 4
        self.buf4len = 0

    def read(self, buf, n):
        """
        :type buf: Destination buffer (List[str])
        :type n: Maximum number of characters to read (int)
        :rtype: The number of characters read (int)
        """
        i = 0
        while i < n:
            if not self.buf4len:
                self.buf4len = read4(self.buf4)
                if not self.buf4len:
                    break
            cur_size = min(n - i, self.buf4len)
            buf[i:] = self.buf4[:cur_size]
            i += cur_size
            self.buf4[:self.buf4len - cur_size] = self.buf4[cur_size : self.buf4len]
            self.buf4len = self.buf4len - cur_size

        return i

File: Java/158__Read_N_Characters_Given_Read4_II-Call_multiple_times.java
/**
 * The read4 API is defined in the parent class Reader4.
 *     int read4(char[] buf);
 */
public class Solution extends Reader4 {
    /**
     * @param buf Destination buffer
     * @param n   Number of characters to read
     * @return    The number of actual characters read
     */
    private int buf4_idx = 0;
    private int currentLengh = 0;
    private char[] buf4;

    public int read(char[] buf, int n) {

        if (buf4 == null) {
            buf4 = new char[4];
            buf4_idx = 0;
            currentLengh = read4(buf4);
        }
        int cur_idx = 0;

        while (cur_idx < n) {
            for (; buf4_idx < Math.min(currentLengh, n - cur_idx + buf4_idx); cur_idx++, buf4_idx++) {
                buf[cur_idx] = buf4[buf4_idx];
            }
            if (cur_idx == n) {
                return n;
            }
            if (currentLengh < 4) {
                return cur_idx;
            }
            currentLengh = read4(buf4);
            buf4_idx = 0;
        }
        return cur_idx;

    }
}
